src/wrapper/jnienv.rs

```diff
@@ -1973,6 +1974,43 @@ impl<'a> JNIEnv<'a> {
         jni_error_code_to_result(res)
     }
 
+    /// Return a tuple with a pointer to elements of the given Java array as first element.
+    /// The tuple's second element indicates if the pointed-to array is a copy or not.
+    ///
+    /// The result is valid until the corresponding release_array_elements() function is
+    /// called. Since the returned array may be a copy of the Java array, changes made to the
+    /// returned array will not necessarily be reflected in the original array until
+    /// release_array_elements() is called.
+    ///
+    /// See also [`release_array_elements`](struct.JNIEnv.html#method.release_array_elements)
+    pub fn get_array_elements<T>(&self, array: jarray) -> Result<(*mut T, bool)> {
```

Comment:
Is such method easier to use in client code than a specialized? Do you have to specify the type as 
```
foo: (*mut u8, bool) = get_array_elements(array)?;
```
?

Reply:
You can still do something like
```
let (ptr, is_copy): (*mut jbyte, _) = env.get_array_elements(java_array)?;
```
for some "clarity" and robustness. But yes, not the best user experience here.

Reply:
**Update**:
An clearer alternative, using explicit type annotation:
```
let (ptr, is_copy) = env.get_array_elements::<jbyte>(java_array)?;
```

Reply:
Thanks, yeah, both options seem more verbose than a specialized method `get_byte_array_elements`. One possible benefit is that it can cast JNI types to Rust types (e.g., jbyte -> u8), though I don't know if we shall support that.

Reply:
> it can cast JNI types to Rust types (e.g., jbyte -> u8), though I don't know if we shall support that.

That always happens. "JNI types" are just "types aliases" of Rust types (see [jni-sys](https://github.com/sfackler/rust-jni-sys/blob/550c6da1d1e6c86761bc9ef8e9fe5ec1ceaa9e78/src/lib.rs#L10)). Even with the specialized functions, if you take a look at the result's type, it's always a Rust type.
And as you can see, nothing prevents you from using the "JNI types".

My opinion is that this second syntax in particular, based on turbofish, is good / clear enough. Another benefit is implicit type resolution, like in [this usage in jnienv.rs](https://github.com/maurolacy/jni-rs/blob/6cb87706730388c5b64b77b90f6b2237e621126e/src/wrapper/jnienv.rs#L2088).

The issue here, in my opinion, is more one of low-level implementation... we can save some code repetition and methods but, do we want to do a match on type name as part of these functions?

And moreover: Is there a better way to do it (using traits by example), that we are missing? I for one am not happy with converting the type to a string, and doing a match on that.

Reply:
Regarding performance, I realize now that despite its complexity, **all of this will be resolved at compile time**; because types are known / resolved at compile time! (this is obvious, I know; it just took me some time to realize here).

I've implemented a variant using traits but, it's not worth it. I now. also have a clunky impl based on `TypeId::of::<T>()`; but I think using `type_name()` with a `match` is better / clearer.

So, that's it! I don't think there is a better way to do it than the current one. It is for you to decide its worth, and if you want those constructs in the lib.